### DISCUSSION TITLE
How to install the Propel plugin

### DISCUSSION THREAD

**Discussion by DBA-Z21A** [FriendsOfSymfony1/symfony1#333] [Q&A]
This text was created using automatic translation.
Please forgive me if there are any grammatical errors.

I'm very happy that a project like this exists.
A big thank you to everyone involved in the project.

I would like to have a system written in Symfony 1.1, currently running on centOS7, run on PHP 8.0.3.
That system uses Propel for the ORM, so I created a new FriendsOfSymfony1 project as shown below and tried to install the Propel plugin, but it didn't work.

php vendor/friendsofsymfony1/symfony1/data/bin/symfony generate:project SymfonyTestProject --orm=none
(New project generation completed successfully)

vendor/bin/symfony plugin:install sfPropelORMPlugin

>> plugin    installing plugin "sfPropelORMPlugin"
>> sfPearFrontendPlugin Attempting to discover channel "pear.symfony-project.com"...
>> sfPearFrontendPlugin Attempting fallback to https instead of http on channel
>> sfPearFrontendPlugin "pear.symfony-project.com"...

  Unable to register channel "pear.symfony-project.com" (use --force-license to force installation)

vendor/bin/symfony plugin:install sfPropelORMPlugin --force-license
>> plugin    installing plugin "sfPropelORMPlugin"
>> sfPearFrontendPlugin Attempting to discover channel "pear.symfony-project.com"...
>> sfPearFrontendPlugin Attempting fallback to https instead of http on channel
>> sfPearFrontendPlugin "pear.symfony-project.com"...

  Unable to register channel "pear.symfony-project.com"

What should I do to create a new project using Propel?
Any advice would be greatly appreciated.

**Comment by mkopinsky** [comment]
You should not be creating new projects in symfony1. This project exists to provide support for existing applications which are still using symfony1, but new applications should be created in Symfony7 with Doctrine ORM 2 or 3.

**Comment by connorhu** [comment]
> You should not be creating new projects in symfony1. This project exists to provide support for existing applications which are still using symfony1, but new applications should be created in Symfony7 with Doctrine ORM 2 or 3.

I think it's a translation error. He wants to make an existing system work. But correct me if I'm wrong.

**Comment by connorhu** [comment]
@DBA-Z21A If you have an existing system, I think it's easier to update it than to build a completely new skeleton.
These legacy repo's are less testing for code generation, as everyone is upgrading their existing system, so there may be blind spots in the functionality.

**Comment by mkopinsky** [comment]
OK I see, he's trying to port the sf1.1+propel project to use fos1 (i.e. sf1.6), and doing so by trying to create a new project, installing propel, and then copying over the code. I agree with your second comment - he should be updating sf1 in the original project from 1.1 to 1.6 rather than trying to create a new project. I don't know how propel is typically installed, but the symfony pear channels have been down for years, so installing from pear isn't going to work.

**Comment by DBA-Z21A** [comment]
Thank you @connorhu, @mkopinsky.
You are right.
I want to create an empty FriendsOfSymfony1+Propel project and copy the symfony1.1+Propel code to it.
Of course, I don't think that a simple copy will work perfectly!
I understand that various modifications are necessary.
But right now I don't know any other way to upgrade from symfony1.1 to FriendsOfSymfony1.

>the symfony pear channels have been down for years, so installing from pear isn't going to work.

Thank you for your valuable information.
If anyone knows how to install the Propel plugin, please advise.

**Comment by connorhu** [comment]
@DBA-Z21A There are update instructions in the wiki. It's pretty basic, but it contains important informations.
To install propel, you first need to find a proper fork and install it with composer (you will have problems with propel in the future, so be prepared: https://github.com/FriendsOfSymfony1/symfony1/issues/288#issuecomment-1902757086 ). You can install it from the official repo:
`composer require propel/sf-propel-o-r-m-plugin "dev-master"`

If this is done and you have done the parts of the wiki description that are relevant to you, you should be ahead.

**Comment by DBA-Z21A** [comment]
thank you!
I was able to confirm that the files were extracted in the "plugins/sfPropelORMPlugin" directory.

Model creation commands such as "symfony propel:build-model" do not work yet, but I will research how to use them a little more.
Is this the wiki URL?
https://github.com/FriendsOfSymfony1/symfony1/wiki/Migrations
I will read the description about Doctrine carefully and see if I can apply it to Propel.

Thank you for letting me know that there is a problem with Propel Plugin+PHP8.
I will also consider using PHP7.
My top priority is dealing with the end of support for centOS7.

Thank you very much for your appropriate and prompt assistance.

**Comment by connorhu** [comment]
@DBA-Z21A I will convert this issue to discussion. Feel free to ask if you stuck somewhere.

**Comment by DBA-Z21A** [comment]
Thank you for yesterday!
I sincerely appreciate everyone's advice.


After searching some websites and installing sfPropelORMPlugin via Composer, I tried modifying "config/ProjectConfiguration.class.php" as below.

```
<?php
require __DIR__ .'/../vendor/autoload.php';

require_once __DIR__.'/../vendor/friendsofsymfony1/symfony1/lib/autoload/sfCoreAutoload.class.php';
sfCoreAutoload::register();

class ProjectConfiguration extends sfProjectConfiguration
{
  public function setup()
  {
    //setup the location for our phing and propel libs
    sfConfig::set('sf_phing_path', sfConfig::get('sf_root_dir').'/vendor/phing/phing/');
    sfConfig::set('sf_propel_path', sfConfig::get('sf_root_dir').'/vendor/propel/propel1/');
    sfConfig::set('sf_propel_generator_path', sfConfig::get('sf_root_dir').'/vendor/propel/generator/lib/');

    $this->enablePlugins('sfPropelORMPlugin');
  }
}


```
However, the following error occurs.

index.php
Fatal error: Uncaught Error: Call to undefined function create_function() in 
(project root)/vendor/propel/propel1/runtime/lib/parser/yaml/sfYamlInline.php:130 Stack trace: #0

php symfony cc
PHP Fatal error: Cannot declare class XMLElement, because the name is already in use in 
(project root)/vendor/propel/propel1/generator/lib/model/XMLElement.php on line 20

Is this an error related to PHP8 that I received advice about yesterday?
Should I try it with PHP7?

**Comment by DBA-Z21A** [comment]
Thank you to everyone who gave me advice!
Today, when I created an AlmaLinux8+PHP7 environment and modified "config/ProjectConfiguration.class.php" as follows, "symfony propel:build-model" started to work successfully.
*Of course, I got an error "schema.yml does not exist"! :-)

```
<?php

//require_once __DIR__.'/../vendor/friendsofsymfony1/symfony1/lib/autoload/sfCoreAutoload.class.php';
//sfCoreAutoload::register();
require_once dirname(__FILE__).'/../vendor/autoload.php';

class ProjectConfiguration extends sfProjectConfiguration
{
  public function setup()
  {
//setup the location for our phing and propel libs
sfConfig::set('sf_phing_path', sfConfig::get('sf_root_dir').'/vendor/phing/phing/');
sfConfig::set('sf_propel_path', sfConfig::get('sf_root_dir').'/vendor/propel/propel1/');
sfConfig::set('sf_propel_generator_path', sfConfig::get('sf_root_dir').'/vendor/propel/generator/lib/');

$this->enablePlugins('sfPropelORMPlugin');
  }
}

```

It's a shame that it doesn't work with PHP8, but I'm still happy that it gave me a few years' grace.
In the meantime, I pray that my client makes the right choice to "rebuild everything."
God, grant wisdom to my client.

**Reply by connorhu** [reply]
There are propel forks that work on 8.1. Now I've started to build my own fork that will bring the original library up to 8.3 compatibility in a traceable way (with strftime deprecation). If I can get it to a usable level, it might make it under the friendsofsymfony1 organization (with sf plugin of course).

**Reply by DBA-Z21A** [reply]
Thank you for the great news!
I would like to try it right away.
>Propel forks that work on 8.1. 

If it works well, nothing makes me happier!!

**Reply by connorhu** [reply]
If you have a working schema config that you can share with me, even privately, you can speed up my work, because then I don't have to figure out a working config.

**Reply by thePanz** [reply]
@connorhu let me know if you would like to have a sfPropel project under the FriendsOfSymfony1 organization in GitHub ;)

**Reply by connorhu** [reply]
@thePanz 3 forks will be created, in total. A sfPropelORMPlugin (https://github.com/propelorm/sfPropelORMPlugin with name friendsofsymfony1/sfpropelormplugin ), a Propel (https://github.com/propelorm/Propel with name friendsofsymfony1/propel) and Phing ( https://github.com/connorhu/phing/tree/2.x/ with name friendsofsymfony1/phing ). Phing is heavily used by propel and not so compatible with php83 so we have to fork and fix some issues. I think that when a repo is ready (phing is considered ready), I simply hand over my fork and we modify what is needed. The Propel project is a bit more complicated as the tests are not GHA compatible and I don't want to get on the docker train. Although, no doubt, that's the easier way.

**Reply by mentalstring** [reply]
Happy to see interest in Propel has most are in the Doctrine camp. Support for php8.3 for Propel would be great.

As I mentioned at https://github.com/FriendsOfSymfony1/symfony1/issues/288#issuecomment-1898159896 there's already some work done for support up php8.2, although in my changes I have opted for dropping the getter syntax that relies on the deprecated `strftime` since I don't use it. I think the existing attempts to replace `strftime` (not just in propel) with some equivalent all have quirks that come from `strftime`'s own inconsistent behavior (which, I think, is why php is deprecating it) and since Propel has support for another syntax (based on `date()`), that's a viable upgrade path.

As I remember, sfYamlParser.php proved a bit tricky but symfony1's own sfYamlParser.php (which has diverged somewhat) has everything working, so I think some changes came from it. Would be a good opportunity to get those two back in sync if a propel fork will live under fos1.

**Reply by connorhu** [reply]
I used to look at your fork to see what still needs to be done. I will try to make the changes easy to follow and create a clean update path. If necessary, I can then explain/document in separate PRs why a particular change made, what it affects.
With yaml, I figured out that we could make a separate package of yaml classes that both repo's could reference.
I don't yet see if it's feasible, but a big item on my wish list is to throw phing out the window.

Strftime is only a problem for hardcore developers, since sfPropelORMPlugin uses `date` format by default. If someone modified it, they might have a problem. For now, I've left a depr. notice in the code if someone is using the `strftime` format, to let them know they're in trouble.

**Reply by thePanz** [reply]
@connorhu @mentalstring the forks have been created, let me know if you have the right permissions to manage them :)

**Reply by connorhu** [reply]
@thePanz Thank you! My question is, how will I transfer all the PRs (where there are important comments about PRs) and commits that are in my propel fork? https://github.com/connorhu/Propel

**Reply by thePanz** [reply]
@connorhu not sure, can you edit the PRs and target the new repo? never did that before :shrug:

**Reply by connorhu** [reply]
@thePanz After all, I can redo the whole commit history.

**Reply by thirsch** [reply]
You could also create a pr from your master/main to here?

**Reply by connorhu** [reply]
@thirsch I regularly delete branches, so I don't think I can create a new PR from the content of a branch. Anyway, I'll do it again. I can actually make a cleaner solution this way.

**Reply by connorhu** [reply]
@thePanz Please create a propel team with prope(l)r rights to the two new repositories. I want to start setting up the Propel repo.

**Reply by thePanz** [reply]
@connorhu will do, in the meantime the 2 repos are up already. Can you check it you have all the needed permission?

**Reply by connorhu** [reply]
@thePanz I don't know what I should have access to, but I can't turn on the issue function, rename the master branch, etc.

**Reply by thePanz** [reply]
@mentalstring @connorhu I created a Propel tram and added the 2 repos into it. You're listed as 'maintainers' now

**Reply by connorhu** [reply]
@thePanz Thank you!
